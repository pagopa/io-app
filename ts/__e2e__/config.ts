import { device } from "detox";
import { Millisecond } from "@pagopa/ts-commons/lib/units";

// 10 seconds seems a lot in development, but lower values are causing false positives
// on the CI environment. Don't touch it if you don't know what you are doing.
export const e2eWaitRenderTimeout = (15 * 1000) as Millisecond;
export const e2ePinChar1 = "1";
export const e2ePinChar2 = "2";
export const e2ePinChar3 = "3";
export const e2ePinChar4 = "4";
export const e2ePinChar5 = "5";
export const e2ePinChar6 = "6";

// We need to extract the argument type because is not exported
type DetoxLaunchAppConfig = Parameters<typeof device.launchApp>[0];

export const launchAppConfig: DetoxLaunchAppConfig = {
  permissions: { notifications: "YES", camera: "YES", photos: "YES" },
  // enforce IT locale because that's how the API are configured
  languageAndLocale: { language: "it_IT", locale: "it_IT" }
};
