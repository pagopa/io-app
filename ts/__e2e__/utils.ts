import {
  e2ePinChar1,
  e2ePinChar2,
  e2ePinChar3,
  e2ePinChar4,
  e2ePinChar5,
  e2ePinChar6,
  e2eWaitRenderTimeout
} from "./config";

const onboardingPinTitleId = "pin-creation-form-title";

/**
 * Complete the login with SPID
 */
export const loginWithSPID = async () => {
  const loginButtonId = "landing-button-login-spid";
  await waitFor(element(by.id(loginButtonId)))
    .toBeVisible()
    .withTimeout(e2eWaitRenderTimeout);

  await element(by.id(loginButtonId)).tap();

  const posteIdpButtonId = "idp-posteid-button";
  await waitFor(element(by.id(posteIdpButtonId)))
    .toBeVisible()
    .withTimeout(e2eWaitRenderTimeout);

  await element(by.id(posteIdpButtonId)).tap();

  // the webview returned by the server has 250ms timeout and reloads automagically

  const shareDataComponentTitleId = "share-data-component-title";
  await waitFor(element(by.id(shareDataComponentTitleId)))
    .toBeVisible()
    .withTimeout(e2eWaitRenderTimeout);

  const shareDataRightButtonId = "share-data-confirm-button";
  await element(by.id(shareDataRightButtonId)).tap();

  await waitFor(element(by.id(onboardingPinTitleId)))
    .toBeVisible()
    .withTimeout(e2eWaitRenderTimeout);

  await createE2EPin();
};

/**
 * This utility should be used during the authentication
 * step in the pin screen. It will leverage the IO Pinpad
 * to insert the pin.
 */
export const insertE2EPin = async () => {
  await element(by.text(e2ePinChar1)).tap();
  await element(by.text(e2ePinChar6)).tap();
  await element(by.text(e2ePinChar2)).tap();
  await element(by.text(e2ePinChar5)).tap();
  await element(by.text(e2ePinChar3)).tap();
  await element(by.text(e2ePinChar4)).tap();
};

/**
 * This utility should be used during the onboarding in the
 * pin creation screen, or in the same screen in the Profile section.
 * It will fill the two inputs with a pin and submit the form also
 * trying to insert wrong data during the process.
 */
export const createE2EPin = async () => {
  const pin =
    e2ePinChar1 +
    e2ePinChar6 +
    e2ePinChar2 +
    e2ePinChar5 +
    e2ePinChar3 +
    e2ePinChar4;
  const wrongPin = "123456";

  const onboardingPinFieldInputId = "PinFieldInput";
  await element(by.id(onboardingPinFieldInputId)).typeText(pin);

  const scrollView = element(by.id("pin-creation-form-scroll-view"));
  await scrollView.scrollTo("bottom");

  const onboardingPinConfirmationFieldId = "PinConfirmationFieldInput";
  await element(by.id(onboardingPinConfirmationFieldId)).typeText(wrongPin);

  const onboardingPinConfirmButtonId = "pin-creation-form-confirm";
  await waitFor(element(by.id(onboardingPinConfirmButtonId)))
    .toBeVisible()
    .withTimeout(e2eWaitRenderTimeout);

  await scrollView.tap();

  await element(by.id(onboardingPinConfirmationFieldId)).clearText();
  await element(by.id(onboardingPinConfirmationFieldId)).typeText(pin);

  await element(by.id(onboardingPinConfirmButtonId)).tap();

  const onboardingNotEnrolledConfirmButtonId = "not-enrolled-biometric-confirm";
  await waitFor(element(by.id(onboardingNotEnrolledConfirmButtonId)))
    .toBeVisible()
    .withTimeout(e2eWaitRenderTimeout);
  await element(by.id(onboardingNotEnrolledConfirmButtonId)).tap();
};

/**
 * Ensures that the login with spid and the unlock code has been inserted
 */
export const ensureLoggedIn = async () => {
  try {
    const identificationModalBodyId = "identification-modal-body";
    await waitFor(element(by.id(identificationModalBodyId)))
      .toBeVisible()
      .withTimeout(e2eWaitRenderTimeout);
    await insertE2EPin();
  } catch {
    await loginWithSPID();
  }
};
