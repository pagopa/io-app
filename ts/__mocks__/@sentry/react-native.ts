export const captureException = jest.fn();
export const captureMessage = jest.fn();
export const captureEvent = jest.fn();
export const setContext = jest.fn();
export const setExtras = jest.fn();
export const setExtra = jest.fn();
export const setTags = jest.fn();
export const setTag = jest.fn();
export const setUser = jest.fn();
export const lastEventId = jest.fn();
export const captureCheckIn = jest.fn();
export const withMonitor = jest.fn();
export const flush = jest.fn().mockResolvedValue(false);
export const close = jest.fn().mockResolvedValue(false);
export const isInitialized = jest.fn().mockReturnValue(false);
export const isEnabled = jest.fn().mockReturnValue(false);
export const addEventProcessor = jest.fn();
export const startSession = jest.fn();
export const endSession = jest.fn();
export const captureSession = jest.fn();
export const ErrorBoundary = jest.fn(({ children }) => children);
