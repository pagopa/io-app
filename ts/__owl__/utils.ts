import { press, toExist } from "react-native-owl";

const loginButtonId = "landing-button-login-spid";
const posteIdpButtonId = "idp-posteid-button";

/**
 * Wait for Login Screen
 */
export const waitForLoginScreen = async () => await toExist(loginButtonId);

/**
 * Wait for SPID IdPs Screen
 */
export const waitForSpidScreen = async () => {
  await waitForLoginScreen();
  await press(loginButtonId);
  await toExist(posteIdpButtonId);
};

/**
 * Login with PosteID
 */
export const loginWithPosteID = async () => {
  await waitForSpidScreen();
  await press(posteIdpButtonId);
  // await toExist("buttonRight");
};
