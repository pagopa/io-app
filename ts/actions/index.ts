/**
 * Aggregates all the actions available to components.
 */

export * from './login'
export * from './user'
