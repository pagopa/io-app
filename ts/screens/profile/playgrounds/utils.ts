const generateRandomNumberString = (length: number) => {
  const characters = "0123456789";

  return Array.from(Array(length).keys())
    .map(() => {
      const randomIndex = Math.floor(Math.random() * characters.length);
      return characters[randomIndex];
    })
    .join("");
};

export const generateValidRandomRptIdString = () => {
  const organizationFiscalCode = generateRandomNumberString(11);
  const paymentNoticeNumber = "002" + generateRandomNumberString(15);

  return `${organizationFiscalCode}${paymentNoticeNumber}`;
};
