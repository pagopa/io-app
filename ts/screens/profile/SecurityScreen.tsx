import React, { FC, useEffect, useState } from "react";
import { List } from "native-base";
import { connect } from "react-redux";
import { Dispatch } from "redux";
import TouchID, { AuthenticationError } from "react-native-touch-id";
import I18n from "../../i18n";
import { GlobalState } from "../../store/reducers/types";
import { getFingerprintSettings } from "../../sagas/startup/checkAcknowledgedFingerprintSaga";
import TopScreenComponent from "../../components/screens/TopScreenComponent";
import { ContextualHelpPropsMarkdown } from "../../components/screens/BaseScreenComponent";
import ListItemComponent from "../../components/screens/ListItemComponent";
import { updatePin } from "../../store/actions/pinset";
import { identificationRequest } from "../../store/actions/identification";
import { shufflePinPadOnPayment } from "../../config";
import { authenticateConfig } from "../../utils/biometric";
import { showToast } from "../../utils/showToast";
import { preferenceFingerprintIsEnabledSaveSuccess } from "../../store/actions/persistedPreferences";
import { useScreenReaderEnabled } from "../../utils/accessibility";
import ScreenContent from "../../components/screens/ScreenContent";

const contextualHelpMarkdown: ContextualHelpPropsMarkdown = {
  title: "profile.preferences.contextualHelpTitle",
  body: "profile.preferences.contextualHelpContent"
};

type Props = ReturnType<typeof mapDispatchToProps> &
  ReturnType<typeof mapStateToProps>;

const SecurityScreen: FC<Props> = ({
  isFingerprintEnabled,
  requestIdentificationAndResetPin,
  setFingerprintPreference
}): React.ReactElement => {
  const isScreenReaderEnabled = useScreenReaderEnabled();
  const [isFingerprintAvailable, setIsFingerprintAvailable] = useState(false);

  useEffect(() => {
    getFingerprintSettings().then(
      biometryTypeOrUnsupportedReason => {
        setIsFingerprintAvailable(
          biometryTypeOrUnsupportedReason !== "UNAVAILABLE" &&
            biometryTypeOrUnsupportedReason !== "NOT_ENROLLED"
        );
      },
      _ => undefined
    );
  }, []);

  // FIXME: Add alert if user refused IOS biometric permission on
  // the first time the app is opened: https://pagopa.atlassian.net/browse/IA-67

  const setBiometricPreference = (biometricPreference: boolean): void => {
    if (biometricPreference) {
      // if user asks to enable biometric then call enable action directly
      setFingerprintPreference(biometricPreference);
      return;
    }
    // if user asks to disable biometric recnognition is required to proceed
    TouchID.authenticate(
      I18n.t("identification.biometric.popup.title"),
      authenticateConfig
    )
      .then(() => setFingerprintPreference(biometricPreference))
      .catch((_: AuthenticationError) =>
        // this toast will be show either if recognition fails (mismatch or user aborts)
        // or if meanwhile user disables biometric recognition in OS settings
        showToast(
          I18n.t(
            "profile.security.list.biometric_recognition.needed_to_disable"
          ),
          "danger"
        )
      );
  };

  const onPressBiometricRecognitionItem = () => {
    if (isScreenReaderEnabled) {
      setBiometricPreference(!isFingerprintEnabled);
    }
  };

  return (
    <TopScreenComponent
      contextualHelpMarkdown={contextualHelpMarkdown}
      faqCategories={["profile", "privacy", "authentication_SPID"]}
      goBack
    >
      <ScreenContent
        title={I18n.t("profile.security.title")}
        subtitle={I18n.t("profile.security.subtitle")}
      >
        <List withContentLateralPadding>
          {/* Ask for verification and reset unlock code */}
          <ListItemComponent
            title={I18n.t("identification.unlockCode.reset.button_short")}
            subTitle={I18n.t("identification.unlockCode.reset.subtitle")}
            onPress={requestIdentificationAndResetPin}
            testID="reset-unlock-code"
          />
          {/* Enable/disable biometric recognition */}
          {isFingerprintAvailable && (
            <ListItemComponent
              title={I18n.t(
                "profile.security.list.biometric_recognition.title"
              )}
              subTitle={I18n.t(
                "profile.security.list.biometric_recognition.subtitle"
              )}
              onSwitchValueChanged={setBiometricPreference}
              switchValue={isFingerprintEnabled}
              isLongPressEnabled
              accessibilityState={{ checked: isFingerprintEnabled }}
              accessibilityRole="switch"
              onPress={onPressBiometricRecognitionItem}
              testID="biometric-recognition"
            />
          )}
        </List>
      </ScreenContent>
    </TopScreenComponent>
  );
};

const mapDispatchToProps = (dispatch: Dispatch) => ({
  requestIdentificationAndResetPin: () => {
    const onSuccess = () => dispatch(updatePin());

    return dispatch(
      identificationRequest(
        true,
        false,
        undefined,
        undefined,
        {
          onSuccess
        },
        shufflePinPadOnPayment
      )
    );
  },
  setFingerprintPreference: (fingerprintPreference: boolean) =>
    dispatch(
      preferenceFingerprintIsEnabledSaveSuccess({
        isFingerprintEnabled: fingerprintPreference
      })
    )
});

const mapStateToProps = (state: GlobalState) => ({
  isFingerprintEnabled: state.persistedPreferences.isFingerprintEnabled
});

export default connect(mapStateToProps, mapDispatchToProps)(SecurityScreen);
