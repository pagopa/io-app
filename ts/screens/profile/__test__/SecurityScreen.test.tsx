import React from "react";
import configureMockStore from "redux-mock-store";
import { NavigationParams } from "react-navigation";
import { fireEvent } from "@testing-library/react-native";
import SecurityScreen from "../SecurityScreen";
import I18n from "../../../i18n";
import { appReducer } from "../../../store/reducers";
import { applicationChangeState } from "../../../store/actions/application";
import { GlobalState } from "../../../store/reducers/types";
import { renderScreenFakeNavRedux } from "../../../utils/testWrapper";
import ROUTES from "../../../navigation/routes";
import { identificationRequest } from "../../../store/actions/identification";
import { shufflePinPadOnPayment } from "../../../config";
import { updatePin } from "../../../store/actions/pinset";

describe("Test SecurityScreen", () => {
  jest.useFakeTimers();
  it("should be not null", () => {
    const { component } = renderComponent();

    expect(component).not.toBeNull();
  });
  it("should render H1 component with title and H4 component with subtitle", () => {
    const { component } = renderComponent();

    expect(component).not.toBeNull();
    expect(
      component.queryByText(I18n.t("profile.security.title"))
    ).not.toBeNull();
    expect(
      component.queryByText(I18n.t("profile.security.subtitle"))
    ).not.toBeNull();
  });
  it("should render ListItemComponent reset unlock code with the right title and subtitle", () => {
    const { component } = renderComponent();

    expect(component).not.toBeNull();
    expect(component.queryByTestId("reset-unlock-code")).not.toBeNull();
    expect(
      component.queryByText(I18n.t("profile.security.title"))
    ).not.toBeNull();
    expect(
      component.queryByText(I18n.t("profile.security.subtitle"))
    ).not.toBeNull();
  });
  it("when press ListItemComponent reset unlock code, should dispatch identificationRequest", () => {
    const { component, store } = renderComponent();

    expect(component).not.toBeNull();
    const listItemComponent = component.getByTestId("reset-unlock-code");
    expect(listItemComponent).not.toBeNull();
    fireEvent.press(listItemComponent);
    store.dispatch(
      identificationRequest(
        true,
        false,
        undefined,
        undefined,
        {
          onSuccess: () => store.dispatch(updatePin())
        },
        shufflePinPadOnPayment
      )
    );
  });
});

const renderComponent = () => {
  const globalState = appReducer(undefined, applicationChangeState("active"));

  const mockStore = configureMockStore<GlobalState>();
  const store: ReturnType<typeof mockStore> = mockStore({
    ...globalState
  } as GlobalState);

  return {
    component: renderScreenFakeNavRedux<GlobalState, NavigationParams>(
      () => <SecurityScreen />,
      ROUTES.PROFILE_SECURITY,
      {},
      store
    ),
    store
  };
};
