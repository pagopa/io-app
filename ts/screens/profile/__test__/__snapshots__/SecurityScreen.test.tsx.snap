// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test SecurityScreen should match snapshot when 'fimsIsHistoryEnabledSelector' returns 'true' 1`] = `
<View
  style={
    {
      "flex": 1,
    }
  }
>
  <RNCSafeAreaProvider
    onInsetsChange={[Function]}
    style={
      [
        {
          "flex": 1,
        },
        undefined,
      ]
    }
  >
    <View
      style={
        [
          {
            "backgroundColor": "#FFFFFF",
            "flex": 1,
          },
          undefined,
        ]
      }
    >
      <View
        collapsable={false}
        pointerEvents="box-none"
        style={
          {
            "zIndex": 1,
          }
        }
      >
        <View
          accessibilityElementsHidden={false}
          importantForAccessibility="auto"
          onLayout={[Function]}
          pointerEvents="box-none"
          style={null}
        >
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              {
                "bottom": 0,
                "left": 0,
                "opacity": 1,
                "position": "absolute",
                "right": 0,
                "top": 0,
                "zIndex": 0,
              }
            }
          >
            <View
              collapsable={false}
              style={
                {
                  "backgroundColor": "#FFFFFF",
                  "borderBottomColor": "rgb(216, 216, 216)",
                  "flex": 1,
                  "shadowColor": "rgb(216, 216, 216)",
                  "shadowOffset": {
                    "height": 0.5,
                    "width": 0,
                  },
                  "shadowOpacity": 0.85,
                  "shadowRadius": 0,
                }
              }
            />
          </View>
          <View
            collapsable={false}
            pointerEvents="box-none"
            style={
              {
                "height": 44,
                "maxHeight": undefined,
                "minHeight": undefined,
                "opacity": undefined,
                "transform": undefined,
              }
            }
          >
            <View
              pointerEvents="none"
              style={
                {
                  "height": 0,
                }
              }
            />
            <View
              pointerEvents="box-none"
              style={
                {
                  "alignItems": "stretch",
                  "flex": 1,
                  "flexDirection": "row",
                }
              }
            >
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "alignItems": "flex-start",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginStart": 0,
                    "opacity": 1,
                  }
                }
              />
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "justifyContent": "center",
                    "marginHorizontal": 16,
                    "maxWidth": 718,
                    "opacity": 1,
                  }
                }
              >
                <Text
                  accessibilityRole="header"
                  aria-level="1"
                  collapsable={false}
                  numberOfLines={1}
                  onLayout={[Function]}
                  style={
                    {
                      "color": "rgb(28, 28, 30)",
                      "fontSize": 17,
                      "fontWeight": "600",
                    }
                  }
                >
                  PROFILE_SECURITY
                </Text>
              </View>
              <View
                collapsable={false}
                pointerEvents="box-none"
                style={
                  {
                    "alignItems": "flex-end",
                    "flexBasis": 0,
                    "flexGrow": 1,
                    "justifyContent": "center",
                    "marginEnd": 0,
                    "opacity": 1,
                  }
                }
              />
            </View>
          </View>
        </View>
      </View>
      <RNSScreenContainer
        onLayout={[Function]}
        style={
          {
            "flex": 1,
          }
        }
      >
        <RNSScreen
          activityState={2}
          collapsable={false}
          gestureResponseDistance={
            {
              "bottom": -1,
              "end": -1,
              "start": -1,
              "top": -1,
            }
          }
          onGestureCancel={[Function]}
          pointerEvents="box-none"
          sheetAllowedDetents="large"
          sheetCornerRadius={-1}
          sheetExpandsWhenScrolledToEdge={true}
          sheetGrabberVisible={false}
          sheetLargestUndimmedDetent="all"
          style={
            {
              "bottom": 0,
              "left": 0,
              "position": "absolute",
              "right": 0,
              "top": 0,
              "zIndex": undefined,
            }
          }
        >
          <View
            collapsable={false}
            style={
              {
                "opacity": 1,
              }
            }
          />
          <View
            accessibilityElementsHidden={false}
            closing={false}
            gestureVelocityImpact={0.3}
            importantForAccessibility="auto"
            onClose={[Function]}
            onGestureBegin={[Function]}
            onGestureCanceled={[Function]}
            onGestureEnd={[Function]}
            onOpen={[Function]}
            onTransition={[Function]}
            pointerEvents="box-none"
            style={
              [
                {
                  "display": "flex",
                  "overflow": undefined,
                },
                {
                  "bottom": 0,
                  "left": 0,
                  "position": "absolute",
                  "right": 0,
                  "top": 0,
                },
              ]
            }
            transitionSpec={
              {
                "close": {
                  "animation": "spring",
                  "config": {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
                "open": {
                  "animation": "spring",
                  "config": {
                    "damping": 500,
                    "mass": 3,
                    "overshootClamping": true,
                    "restDisplacementThreshold": 10,
                    "restSpeedThreshold": 10,
                    "stiffness": 1000,
                  },
                },
              }
            }
          >
            <View
              collapsable={false}
              pointerEvents="box-none"
              style={
                {
                  "flex": 1,
                }
              }
            >
              <View
                collapsable={false}
                handlerTag={5}
                handlerType="PanGestureHandler"
                needsOffscreenAlphaCompositing={false}
                onGestureHandlerEvent={[Function]}
                onGestureHandlerStateChange={[Function]}
                style={
                  {
                    "flex": 1,
                    "transform": [
                      {
                        "translateX": 0,
                      },
                      {
                        "translateX": 0,
                      },
                    ],
                  }
                }
              >
                <View
                  pointerEvents="box-none"
                  style={
                    [
                      {
                        "flex": 1,
                        "overflow": "hidden",
                      },
                      [
                        {
                          "backgroundColor": "#FFFFFF",
                        },
                        undefined,
                      ],
                    ]
                  }
                >
                  <View
                    style={
                      {
                        "flex": 1,
                        "flexDirection": "column-reverse",
                      }
                    }
                  >
                    <View
                      style={
                        {
                          "flex": 1,
                        }
                      }
                    >
                      <RCTScrollView
                        collapsable={false}
                        contentContainerStyle={
                          [
                            {
                              "paddingBottom": 56,
                              "paddingHorizontal": 0,
                            },
                            {},
                          ]
                        }
                        decelerationRate="normal"
                        onScroll={[Function]}
                        scrollEventThrottle={8}
                        snapToEnd={false}
                        snapToOffsets={
                          [
                            0,
                            0,
                          ]
                        }
                        style={{}}
                      >
                        <View>
                          <View
                            accessible={true}
                            onLayout={[Function]}
                            style={
                              {
                                "paddingHorizontal": 24,
                              }
                            }
                          >
                            <Text
                              accessibilityLabel="Security"
                              accessibilityRole="header"
                              allowFontScaling={false}
                              dynamicTypeRamp="title1"
                              maxFontSizeMultiplier={1.25}
                              style={
                                [
                                  {},
                                  {
                                    "color": "#0E0F13",
                                    "fontFamily": "Titillium Sans Pro",
                                    "fontSize": 28,
                                    "fontStyle": "normal",
                                    "fontWeight": "600",
                                    "lineHeight": 40,
                                  },
                                ]
                              }
                            >
                              Security
                            </Text>
                          </View>
                          <View
                            style={
                              {
                                "paddingHorizontal": 24,
                              }
                            }
                          >
                            <View
                              style={
                                {
                                  "height": 16,
                                }
                              }
                            />
                            <Text
                              allowFontScaling={false}
                              dynamicTypeRamp="body"
                              maxFontSizeMultiplier={1.25}
                              style={
                                [
                                  {},
                                  {
                                    "color": "#555C70",
                                    "fontFamily": "Titillium Sans Pro",
                                    "fontSize": 16,
                                    "fontStyle": "normal",
                                    "fontWeight": "400",
                                    "lineHeight": 24,
                                  },
                                ]
                              }
                            >
                              Manage in-app authentication methods.
                            </Text>
                          </View>
                          <View
                            style={
                              {
                                "height": 16,
                              }
                            }
                          />
                          <View
                            style={
                              {
                                "paddingHorizontal": 24,
                              }
                            }
                          >
                            <View
                              accessibilityLabel="Change the unlock code"
                              accessibilityRole="button"
                              accessibilityState={
                                {
                                  "busy": undefined,
                                  "checked": undefined,
                                  "disabled": undefined,
                                  "expanded": undefined,
                                  "selected": undefined,
                                }
                              }
                              accessibilityValue={
                                {
                                  "max": undefined,
                                  "min": undefined,
                                  "now": undefined,
                                  "text": undefined,
                                }
                              }
                              accessible={true}
                              collapsable={false}
                              focusable={true}
                              onBlur={[Function]}
                              onClick={[Function]}
                              onFocus={[Function]}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              testID="reset-unlock-code"
                            >
                              <View
                                style={
                                  [
                                    {
                                      "marginHorizontal": -24,
                                      "paddingHorizontal": 24,
                                      "paddingVertical": 12,
                                    },
                                    {
                                      "backgroundColor": undefined,
                                    },
                                  ]
                                }
                              >
                                <View
                                  style={
                                    [
                                      {
                                        "alignItems": "center",
                                        "flexDirection": "row",
                                        "justifyContent": "space-between",
                                      },
                                      {
                                        "transform": [
                                          {
                                            "scale": undefined,
                                          },
                                        ],
                                      },
                                    ]
                                  }
                                >
                                  <View
                                    style={
                                      {
                                        "flex": 1,
                                      }
                                    }
                                  >
                                    <Text
                                      allowFontScaling={false}
                                      dynamicTypeRamp="headline"
                                      maxFontSizeMultiplier={1.25}
                                      style={
                                        [
                                          {},
                                          {
                                            "color": "#0E0F13",
                                            "fontFamily": "Titillium Sans Pro",
                                            "fontSize": 18,
                                            "fontStyle": "normal",
                                            "fontWeight": "600",
                                            "lineHeight": 25,
                                          },
                                        ]
                                      }
                                    >
                                      Change the unlock code
                                    </Text>
                                    <Text
                                      allowFontScaling={false}
                                      dynamicTypeRamp="footnote"
                                      maxFontSizeMultiplier={1.25}
                                      style={
                                        [
                                          {},
                                          {
                                            "color": "#555C70",
                                            "fontFamily": "Titillium Sans Pro",
                                            "fontSize": 14,
                                            "fontStyle": "normal",
                                            "fontWeight": "400",
                                            "lineHeight": 21,
                                          },
                                        ]
                                      }
                                    >
                                      Set a new 6-digit code.
                                    </Text>
                                  </View>
                                  <RNSVGSvgView
                                    accessibilityElementsHidden={true}
                                    accessibilityLabel=""
                                    accessible={false}
                                    align="xMidYMid"
                                    bbHeight={24}
                                    bbWidth={24}
                                    focusable={false}
                                    height={24}
                                    importantForAccessibility="no-hide-descendants"
                                    meetOrSlice={0}
                                    minX={0}
                                    minY={0}
                                    style={
                                      [
                                        {
                                          "backgroundColor": "transparent",
                                          "borderWidth": 0,
                                        },
                                        {
                                          "color": "#0073E6",
                                        },
                                        {
                                          "flex": 0,
                                          "height": 24,
                                          "width": 24,
                                        },
                                      ]
                                    }
                                    tintColor="#0073E6"
                                    vbHeight={24}
                                    vbWidth={24}
                                    width={24}
                                  >
                                    <RNSVGGroup
                                      fill={
                                        {
                                          "payload": 4278190080,
                                          "type": 0,
                                        }
                                      }
                                    >
                                      <RNSVGPath
                                        clipRule={0}
                                        d="M16.2929 5.29289c.3905-.39052 1.0237-.39052 1.4142 0l6 6.00001c.3905.3905.3905 1.0237 0 1.4142l-6 6c-.3905.3905-1.0237.3905-1.4142 0-.3905-.3905-.3905-1.0237 0-1.4142L21.5858 12l-5.2929-5.29289c-.3905-.39053-.3905-1.02369 0-1.41422Z"
                                        fill={
                                          {
                                            "type": 2,
                                          }
                                        }
                                        fillRule={0}
                                        propList={
                                          [
                                            "fill",
                                            "fillRule",
                                          ]
                                        }
                                      />
                                    </RNSVGGroup>
                                  </RNSVGSvgView>
                                </View>
                              </View>
                            </View>
                            <View
                              style={
                                {
                                  "backgroundColor": "#D2D6E3",
                                  "height": 0.5,
                                }
                              }
                            />
                            <View
                              accessibilityRole="button"
                              accessibilityState={
                                {
                                  "busy": undefined,
                                  "checked": undefined,
                                  "disabled": undefined,
                                  "expanded": undefined,
                                  "selected": undefined,
                                }
                              }
                              accessibilityValue={
                                {
                                  "max": undefined,
                                  "min": undefined,
                                  "now": undefined,
                                  "text": undefined,
                                }
                              }
                              accessible={true}
                              collapsable={false}
                              focusable={true}
                              onBlur={[Function]}
                              onClick={[Function]}
                              onFocus={[Function]}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              testID="fims-history"
                            >
                              <View
                                style={
                                  [
                                    {
                                      "marginHorizontal": -24,
                                      "paddingHorizontal": 24,
                                      "paddingVertical": 12,
                                    },
                                    {
                                      "backgroundColor": undefined,
                                    },
                                  ]
                                }
                              >
                                <View
                                  style={
                                    [
                                      {
                                        "alignItems": "center",
                                        "flexDirection": "row",
                                        "justifyContent": "space-between",
                                      },
                                      {
                                        "transform": [
                                          {
                                            "scale": undefined,
                                          },
                                        ],
                                      },
                                    ]
                                  }
                                >
                                  <View
                                    style={
                                      {
                                        "flex": 1,
                                      }
                                    }
                                  >
                                    <Text
                                      allowFontScaling={false}
                                      dynamicTypeRamp="headline"
                                      maxFontSizeMultiplier={1.25}
                                      style={
                                        [
                                          {},
                                          {
                                            "color": "#0E0F13",
                                            "fontFamily": "Titillium Sans Pro",
                                            "fontSize": 18,
                                            "fontStyle": "normal",
                                            "fontWeight": "600",
                                            "lineHeight": 25,
                                          },
                                        ]
                                      }
                                    >
                                      Controlla i tuoi accessi
                                    </Text>
                                    <Text
                                      allowFontScaling={false}
                                      dynamicTypeRamp="footnote"
                                      maxFontSizeMultiplier={1.25}
                                      style={
                                        [
                                          {},
                                          {
                                            "color": "#555C70",
                                            "fontFamily": "Titillium Sans Pro",
                                            "fontSize": 14,
                                            "fontStyle": "normal",
                                            "fontWeight": "400",
                                            "lineHeight": 21,
                                          },
                                        ]
                                      }
                                    >
                                      Consulta la cronologia degli accessi con IO
                                    </Text>
                                  </View>
                                  <RNSVGSvgView
                                    accessibilityElementsHidden={true}
                                    accessibilityLabel=""
                                    accessible={false}
                                    align="xMidYMid"
                                    bbHeight={24}
                                    bbWidth={24}
                                    focusable={false}
                                    height={24}
                                    importantForAccessibility="no-hide-descendants"
                                    meetOrSlice={0}
                                    minX={0}
                                    minY={0}
                                    style={
                                      [
                                        {
                                          "backgroundColor": "transparent",
                                          "borderWidth": 0,
                                        },
                                        {
                                          "color": "#0073E6",
                                        },
                                        {
                                          "flex": 0,
                                          "height": 24,
                                          "width": 24,
                                        },
                                      ]
                                    }
                                    tintColor="#0073E6"
                                    vbHeight={24}
                                    vbWidth={24}
                                    width={24}
                                  >
                                    <RNSVGGroup
                                      fill={
                                        {
                                          "payload": 4278190080,
                                          "type": 0,
                                        }
                                      }
                                    >
                                      <RNSVGPath
                                        clipRule={0}
                                        d="M16.2929 5.29289c.3905-.39052 1.0237-.39052 1.4142 0l6 6.00001c.3905.3905.3905 1.0237 0 1.4142l-6 6c-.3905.3905-1.0237.3905-1.4142 0-.3905-.3905-.3905-1.0237 0-1.4142L21.5858 12l-5.2929-5.29289c-.3905-.39053-.3905-1.02369 0-1.41422Z"
                                        fill={
                                          {
                                            "type": 2,
                                          }
                                        }
                                        fillRule={0}
                                        propList={
                                          [
                                            "fill",
                                            "fillRule",
                                          ]
                                        }
                                      />
                                    </RNSVGGroup>
                                  </RNSVGSvgView>
                                </View>
                              </View>
                            </View>
                          </View>
                        </View>
                      </RCTScrollView>
                    </View>
                  </View>
                </View>
              </View>
            </View>
          </View>
        </RNSScreen>
      </RNSScreenContainer>
    </View>
  </RNCSafeAreaProvider>
</View>
`;
